xonsh/environ.py

```diff
@@ -7,62 +7,188 @@
 import builtins
 import platform
 import subprocess
+from datetime import datetime
+from functools import partial
 from warnings import warn
 
 from xonsh.tools import TERM_COLORS
 
-def current_branch(cwd=None):
-    """Gets the branch for a current working directory. Returns None
-    if the cwd is not a repository.  This currently only works for git, 
-    bust should be extended in the future.
-    """
-    branch = None
-    cwd = os.getcwd() if cwd is None else cwd
+class PromptFormatter(dict):
+    '''
+    Base class that implements utilities for PromptFormatters.
 
-    # step out completely if git is not installed
-    try:
-        binary_location = subprocess.check_output(['which', 'git'], cwd=cwd,
-                                    stderr=subprocess.PIPE,
-                                    universal_newlines=True)
-        if not binary_location:
-            return branch
-    except subprocess.CalledProcessError:
-        return branch
+    You should read the docs for this class to learn how to write your own
+    Prompt Formatter.  However, you most likely want to inherit from
+    DefaultPromptFormatter if you are writing a custom formatter.
+
+    Custom Prompt Formatters will need to implement their own __init__()
+    method that adds entries to several dicts depending on the needs of the
+    variables they add.  The Formatter may also implement methods to implement
+    those variables (although very simple variables can be written with
+    pre-existing functions.)
+
+    Example:
+
+    def __init__(self, *args, **kwargs):
+        super(PromptFormatter, self).__init__(*args, **kwargs)
+
+        # Add a variable that's just a static string.  Useful for constants
+        # that you can't remember:
+        self['bell'] = '\x07'
+
+        # To add a variable that only needs to be computed once add
+        # the function to be called to self.  This can be used for any
+        # variable that won't change while this shell instance is run.  It is
+        # especially useful if the variable takes a long time to compute and
+        # is infrequently used by users:
+        self['login_time'] = self.login_time
+
+        # To add a variable that needs to be computed every time the prompt is
+        # printed simply add a function to self and then add the variable name
+        # to the self._run_every set.
+        self['time24'] = functools.partial(time.strftime, '%H:%M:%S')
+        self._run_every.add('time24')
+
+    def login_time(self):
+        who_out = subprocess.check_output(['who', '-m'])
+        return ' '.join(who_out.decode().split()[2:4])
+    '''
+
+    def __init__(self, *args, **kwargs):
+
+        super(PromptFormatter, self).__init__(*args, **kwargs)
+        self._run_every = set()
```

Comment:
I understand.  What I was trying to say was that I don't think that there is a good reason for this to be private/protected.  Rather, this could conceivably be something that a user wants to change and therefore should be public.  

This is less true of env, where there are many other options for users to manipulate the the value.